File: android/app/src/main/java/app/releaf/mobile/features/home/HomeDashboardViewModel.kt
/*
 * HomeDashboardViewModel.kt
 *
 * Powers the redesigned Home screen: a compact dashboard of
 * notebook + notepad stat cards.
 *
 * Everything is Room-backed — no drive-fake data. Notebook counts
 * come off [NotebookRepository], notepad entries off
 * [NotepadRepository]. Today's notepad count is filtered locally
 * against the device's calendar day.
 */

package app.releaf.mobile.features.home

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import app.releaf.mobile.ReleafApp
import app.releaf.mobile.auth.GoogleAuthSession
import app.releaf.mobile.data.domain.CaptureCountsByMode
import app.releaf.mobile.data.notebook.ChapterRepository
import app.releaf.mobile.data.notebook.NotebookCountRow
import app.releaf.mobile.data.notebook.NotebookEntity
import app.releaf.mobile.data.notebook.NotebookRepository
import app.releaf.mobile.data.notebook.PageRepository
import app.releaf.mobile.data.notebook.parseAttachments
import app.releaf.mobile.data.notebook.parseContacts
import app.releaf.mobile.data.notebook.parseLocations
import app.releaf.mobile.data.notebook.parseTodos
import app.releaf.mobile.data.notepad.NotepadEntry
import app.releaf.mobile.data.notepad.NotepadRepository
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

/** One notebook row on the Home dashboard's Recent list. */
data class NotebookListItem(
    val id: String,
    val title: String,
    val chapterCount: Int,
    val updatedLabel: String,
    val isArchived: Boolean,
)

/** One notepad row on the Home dashboard's Recent list. */
data class NotepadListItem(
    val id: String,
    val title: String,
    val entryDateLabel: String,
    val photoCount: Int,
)

data class HomeDashboardUiState(
    val isLoading: Boolean = true,
    val totalNotebooks: Int = 0,
    val activeNotebooks: Int = 0,
    val archivedNotebooks: Int = 0,
    val totalNotepadEntries: Int = 0,
    val todayNotepadCount: Int = 0,
    val totalNotepadPhotos: Int = 0,
    val totalNotepadScans: Int = 0,
    val totalNotepadVoice: Int = 0,
    val totalNotepadContacts: Int = 0,
    val totalNotepadLocations: Int = 0,
    /** Sum of every attachment-style capture across notepad entries
     *  (photos + scans + voice + contacts + locations). Shown as the
     *  "CAPTURES" stat on the Home library card — entry count and
     *  open todos are rendered as their own lines. */
    val totalNotepadCaptures: Int = 0,
    val openNotepadTodos: Int = 0,
    val recentNotebooks: List<NotebookListItem> = emptyList(),
    val recentNotepadEntries: List<NotepadListItem> = emptyList(),
    val captureCounts: CaptureCountsByMode = CaptureCountsByMode.EMPTY,
)

class HomeDashboardViewModel(
    application: Application,
    private val session: GoogleAuthSession,
    private val notebookRepository: NotebookRepository,
    chapterRepository: ChapterRepository,
    pageRepository: PageRepository,
    private val notepadRepository: NotepadRepository,
) : AndroidViewModel(application) {

    private val activeFlow  = notebookRepository.observeActive()
    private val archivedFlow = notebookRepository.observeArchived()
    private val chapterCountsFlow: Flow<Map<String, Int>> =
        chapterRepository.observeChapterCounts().mapToCountMap()
    private val pageCountsFlow: Flow<Map<String, Int>> =
        pageRepository.observePageCountsByNotebook().mapToCountMap()
    private val notepadFlow = notepadRepository.observeActive(session.userId)

    val state: StateFlow<HomeDashboardUiState> = combine(
        activeFlow, archivedFlow, chapterCountsFlow, pageCountsFlow, notepadFlow,
    ) { active, archived, chapterCounts, pageCounts, notepad ->
        val today = LocalDate.now().toString()
        val todayCount = notepad.count { it.entryDate == today }

        val mostRecentlyUpdated = (active + archived)
            .sortedByDescending { it.updatedAt }
            .take(RECENT_NOTEBOOK_LIMIT)
            .map { entity ->
                NotebookListItem(
                    id           = entity.id,
                    title        = entity.title.ifBlank { "Untitled notebook" },
                    chapterCount = chapterCounts[entity.id] ?: 0,
                    updatedLabel = relativeShort(parseIso(entity.updatedAt)),
                    isArchived   = entity.archivedAt != null,
                )
            }

        val recentEntries = notepad
            .sortedByDescending { it.updatedAt }
            .take(RECENT_NOTEPAD_LIMIT)
            .map { entry ->
                val photoCount = runCatching { entry.attachments.parseAttachments() }
                    .getOrDefault(emptyList())
                    .count { it.type == "photo" }
                NotepadListItem(
                    id             = entry.id,
                    title          = displayTitle(entry),
                    entryDateLabel = humanEntryDate(entry.entryDate),
                    photoCount     = photoCount,
                )
            }

        // Summations across every notepad entry's page features:
        // photos / scans / voice from `attachments` JSON split by
        // `type`, open todos from `todos`, and contacts from
        // `contacts`. In-memory reductions — no extra DB hit.
        val parsedAttachments = notepad.map { entry ->
            runCatching { entry.attachments.parseAttachments() }.getOrDefault(emptyList())
        }
        val totalNotepadPhotos   = parsedAttachments.sumOf { list -> list.count { it.type == "photo" } }
        val totalNotepadScans    = parsedAttachments.sumOf { list -> list.count { it.type == "scan"  } }
        val totalNotepadVoice    = parsedAttachments.sumOf { list -> list.count { it.type == "voice" } }
        val totalNotepadContacts = notepad.sumOf { entry ->
            runCatching { entry.contacts.parseContacts() }.getOrDefault(emptyList()).size
        }
        val totalNotepadLocations = notepad.sumOf { entry ->
            runCatching { entry.locations.parseLocations() }.getOrDefault(emptyList()).size
        }
        val openNotepadTodos = notepad.sumOf { entry ->
            runCatching { entry.todos.parseTodos() }
                .getOrDefault(emptyList())
                .count { !it.done }
        }

        // Aggregate capture counts for the Home-tab trees-saved hero
        // by summing notebook-side + notepad-side contributions for
        // each mode. Notebook page count stands in as "notes" until
        // the captures-table migration ships with per-page breakdowns
        // of photos / scans / voice / contacts / locations.
        val captureCountsCombined = CaptureCountsByMode(
            notes     = pageCounts.values.sum() + notepad.size,
            photos    = totalNotepadPhotos,
            scans     = totalNotepadScans,
            voice     = totalNotepadVoice,
            contacts  = totalNotepadContacts,
            locations = totalNotepadLocations,
        )

        HomeDashboardUiState(
            isLoading            = false,
            totalNotebooks       = active.size + archived.size,
            activeNotebooks      = active.size,
            archivedNotebooks    = archived.size,
            totalNotepadEntries  = notepad.size,
            todayNotepadCount    = todayCount,
            totalNotepadPhotos    = totalNotepadPhotos,
            totalNotepadScans     = totalNotepadScans,
            totalNotepadVoice     = totalNotepadVoice,
            totalNotepadContacts  = totalNotepadContacts,
            totalNotepadLocations = totalNotepadLocations,
            totalNotepadCaptures  = totalNotepadPhotos + totalNotepadScans +
                                    totalNotepadVoice + totalNotepadContacts +
                                    totalNotepadLocations,
            openNotepadTodos      = openNotepadTodos,
            recentNotebooks      = mostRecentlyUpdated,
            recentNotepadEntries = recentEntries,
            captureCounts        = captureCountsCombined,
        )
    }.stateIn(
        scope        = viewModelScope,
        started      = SharingStarted.WhileSubscribed(5_000),
        initialValue = HomeDashboardUiState(),
    )

    /** Create a new notebook. The row appears via the observer stream. */
    fun createNotebook(onCreated: (String) -> Unit = {}) {
        viewModelScope.launch {
            val created = notebookRepository.createNotebook(title = "Untitled notebook")
            onCreated(created.id)
        }
    }

    /** Create a new notepad entry for today. */
    fun createNotepadEntry(onCreated: (String) -> Unit = {}) {
        viewModelScope.launch {
            val entry = notepadRepository.create(
                userId = session.userId,
                entryDate = LocalDate.now().toString(),
                title = null,
                notes = "",
            )
            onCreated(entry.id)
        }
    }

    companion object {
        private const val RECENT_NOTEBOOK_LIMIT = 2
        private const val RECENT_NOTEPAD_LIMIT = 2

        fun factory(session: GoogleAuthSession): ViewModelProvider.Factory = viewModelFactory {
            initializer {
                val app = this[ViewModelProvider.AndroidViewModelFactory.APPLICATION_KEY] as ReleafApp
                HomeDashboardViewModel(
                    application        = app,
                    session            = session,
                    notebookRepository = app.notebookRepository,
                    chapterRepository  = app.chapterRepository,
                    pageRepository     = app.pageRepository,
                    notepadRepository  = app.notepadRepository,
                )
            }
        }
    }
}

// ---------- helpers ----------

private fun Flow<List<NotebookCountRow>>.mapToCountMap(): Flow<Map<String, Int>> =
    map { rows -> rows.associateBy({ it.notebookId }, { it.count }) }

private fun parseIso(iso: String): Instant =
    runCatching { Instant.parse(iso) }.getOrDefault(Instant.EPOCH)

private fun displayTitle(entry: NotepadEntry): String {
    val t = entry.title?.trim().orEmpty()
    if (t.isNotEmpty()) return t
    val firstLine = entry.notes.lineSequence()
        .map { it.trim() }
        .firstOrNull { it.isNotEmpty() }
    return firstLine?.take(80) ?: "Untitled entry"
}

private fun humanEntryDate(isoDate: String): String {
    return runCatching {
        val date = LocalDate.parse(isoDate)
        DateTimeFormatter.ofPattern("MMM d, yyyy").format(date)
    }.getOrDefault(isoDate)
}

File: apps/releaf/ios/Releaf/Features/Home/HomeDashboardViewModel.swift
/*
 * HomeDashboardViewModel.swift
 *
 * Powers the redesigned Home screen: a compact dashboard of
 * notebook + notepad stat cards.
 * All reads are Room/GRDB-backed — no drive-fake data — so the
 * Home dashboard always reflects what the user has actually
 * captured.
 */

import Foundation
import ReleafData

/// One notebook row on the Home dashboard's Recent list.
public struct NotebookListItem: Identifiable, Equatable {
    public let id: String
    public let title: String
    public let chapterCount: Int
    public let updatedLabel: String
    public let isArchived: Bool
}

/// One notepad row on the Home dashboard's Recent list.
public struct NotepadListItem: Identifiable, Equatable {
    public let id: String
    public let title: String
    public let entryDateLabel: String
    public let photoCount: Int
}

public struct HomeDashboardState: Equatable {
    public var isLoading: Bool
    public var totalNotebooks: Int
    public var activeNotebooks: Int
    public var archivedNotebooks: Int
    public var totalNotepadEntries: Int
    public var todayNotepadCount: Int
    public var totalNotepadPhotos: Int
    public var totalNotepadScans: Int
    public var totalNotepadVoice: Int
    public var totalNotepadContacts: Int
    public var totalNotepadLocations: Int
    public var openNotepadTodos: Int
    public var recentNotebooks: [NotebookListItem]
    public var recentNotepadEntries: [NotepadListItem]
    public var captureCounts: CaptureCountsByMode

    public static let initial = HomeDashboardState(
        isLoading: true,
        totalNotebooks: 0,
        activeNotebooks: 0,
        archivedNotebooks: 0,
        totalNotepadEntries: 0,
        todayNotepadCount: 0,
        totalNotepadPhotos: 0,
        totalNotepadScans: 0,
        totalNotepadVoice: 0,
        totalNotepadContacts: 0,
        totalNotepadLocations: 0,
        openNotepadTodos: 0,
        recentNotebooks: [],
        recentNotepadEntries: [],
        captureCounts: .empty
    )
}

@MainActor
public final class HomeDashboardViewModel: ObservableObject {

    @Published public private(set) var state: HomeDashboardState = .initial

    private let userId: String
    private let notebookRepository: NotebookRepository
    private let notepadRepository: NotepadRepository

    private var tasks: [Task<Void, Never>] = []
    private var latestShelves: [ShelfRecord] = []
    private var latestNotepad: [NotepadEntry] = []

    public init(
        userId: String,
        notebookRepository: NotebookRepository = NotebookRepository(),
        notepadRepository: NotepadRepository = NotepadRepository()
    ) {
        self.userId = userId
        self.notebookRepository = notebookRepository
        self.notepadRepository = notepadRepository
    }

    public func start() {
        stop()
        tasks.append(Task { [weak self, notebookRepository] in
            guard let self else { return }
            do {
                for try await value in notebookRepository.observeShelves() {
                    self.latestShelves = value
                    self.recompute()
                }
            } catch {}
        })
        tasks.append(Task { [weak self, notepadRepository, userId] in
            guard let self else { return }
            do {
                for try await value in notepadRepository.observeActive(userId: userId) {
                    self.latestNotepad = value
                    self.recompute()
                }
            } catch {}
        })
    }

    public func stop() {
        tasks.forEach { $0.cancel() }
        tasks.removeAll()
    }

    deinit { tasks.forEach { $0.cancel() } }

    // MARK: - Actions

    public func createNotebook(onCreated: @escaping (String) -> Void = { _ in }) {
        Task { [notebookRepository] in
            if let entity = try? await notebookRepository.createNotebook(title: "Untitled notebook") {
                await MainActor.run { onCreated(entity.id) }
            }
        }
    }

    public func createNotepadEntry(onCreated: @escaping (String) -> Void = { _ in }) {
        Task { [notepadRepository, userId] in
            if let entry = try? await notepadRepository.create(userId: userId, title: nil, notes: "") {
                await MainActor.run { onCreated(entry.id) }
            }
        }
    }

    // MARK: - Recompute

    private func recompute() {
        let total = latestShelves.count
        let active = latestShelves.count   // iOS schema has no archivedAt today
        let todayStr = Self.todayLocalString()
        let todayCount = latestNotepad.filter { $0.entryDate == todayStr }.count

        let recentNotebooks = latestShelves
            .sorted { ($0.notebook.updatedAt) > ($1.notebook.updatedAt) }
            .prefix(2)
            .map { record in
                NotebookListItem(
                    id: record.notebook.id,
                    title: record.notebook.title.nonEmpty ?? "Untitled notebook",
                    chapterCount: record.chapterCount,
                    updatedLabel: Self.relativeShort(
                        isoDate: record.notebook.updatedAt,
                        now: Date()
                    ),
                    isArchived: false
                )
            }

        let recentEntries = latestNotepad
            .sorted { $0.updatedAt > $1.updatedAt }
            .prefix(2)
            .map { entry -> NotepadListItem in
                let photoCount = entry.attachments.parseAttachments()
                    .filter { $0.type == Attachment.typePhoto }.count
                return NotepadListItem(
                    id: entry.id,
                    title: Self.displayTitle(for: entry),
                    entryDateLabel: Self.humanEntryDate(entry.entryDate),
                    photoCount: photoCount
                )
            }

        // Aggregate notepad-page feature stats across every entry:
        // photos / scans / voice come from the `attachments` JSON
        // discriminated by `type`, todos from `todos` JSON (open
        // only), contacts from the `contacts` JSON column. All are
        // in-memory reductions — no extra DB hit. Declared here
        // (above the `captureCounts` initializer) so the trees-saved
        // hero numbers can read them without a forward reference.
        let parsedAttachments = latestNotepad.map { $0.attachments.parseAttachments() }
        let totalNotepadPhotos   = parsedAttachments.reduce(0) { $0 + $1.filter { $0.type == Attachment.typePhoto }.count }
        let totalNotepadScans    = parsedAttachments.reduce(0) { $0 + $1.filter { $0.type == Attachment.typeScan  }.count }
        let totalNotepadVoice    = parsedAttachments.reduce(0) { $0 + $1.filter { $0.type == Attachment.typeVoice }.count }
        let totalNotepadContacts = latestNotepad.reduce(0) { acc, entry in
            acc + entry.contacts.parseContacts().count
        }
        let totalNotepadLocations = latestNotepad.reduce(0) { acc, entry in
            acc + entry.locations.parseLocations().count
        }
        let openNotepadTodos = latestNotepad.reduce(0) { acc, entry in
            acc + entry.todos.parseTodos().filter { !$0.done }.count
        }

        // Aggregate capture counts for the Home-tab trees-saved hero
        // by summing notebook-side + notepad-side contributions for
        // each mode. Notebook page count stands in as "notes" until
        // the captures-table migration lands with per-page breakdowns
        // of photos / scans / voice / contacts / locations.
        let captureCounts = CaptureCountsByMode(
            notes:     latestShelves.reduce(0) { $0 + $1.pageCount } + latestNotepad.count,
            photos:    totalNotepadPhotos,
            scans:     totalNotepadScans,
            voice:     totalNotepadVoice,
            contacts:  totalNotepadContacts,
            locations: totalNotepadLocations
        )

        state = HomeDashboardState(
            isLoading: false,
            totalNotebooks: total,
            activeNotebooks: active,
            archivedNotebooks: 0,
            totalNotepadEntries: latestNotepad.count,
            todayNotepadCount: todayCount,
            totalNotepadPhotos: totalNotepadPhotos,
            totalNotepadScans: totalNotepadScans,
            totalNotepadVoice: totalNotepadVoice,
            totalNotepadContacts: totalNotepadContacts,
            totalNotepadLocations: totalNotepadLocations,
            openNotepadTodos: openNotepadTodos,
            recentNotebooks: Array(recentNotebooks),
            recentNotepadEntries: Array(recentEntries),
            captureCounts: captureCounts
        )
    }

    // MARK: - Helpers

    private static func todayLocalString() -> String {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        return f.string(from: Date())
    }

    private static func displayTitle(for entry: NotepadEntry) -> String {
        if let t = entry.title?.trimmingCharacters(in: .whitespacesAndNewlines), !t.isEmpty {
            return t
        }
        let firstLine = entry.notes
            .split(separator: "\n")
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .first(where: { !$0.isEmpty })
        if let firstLine { return String(firstLine.prefix(80)) }
        return "Untitled entry"
    }

    private static func humanEntryDate(_ iso: String) -> String {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        guard let date = f.date(from: iso) else { return iso }
        let out = DateFormatter()
        out.dateFormat = "MMM d, yyyy"
        return out.string(from: date)
    }

    /// Best-effort "updated 4 days ago" / "2h ago". Parses ISO-8601 from
    /// the entity's `updatedAt` column.
    private static func relativeShort(isoDate: String, now: Date) -> String {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let parsed = f.date(from: isoDate) ?? {
            f.formatOptions = [.withInternetDateTime]
            return f.date(from: isoDate)
        }()
        guard let date = parsed else { return isoDate }
        let delta = max(0, Int(now.timeIntervalSince(date)))
        if delta < 60         { return "just now" }
        if delta < 3600       { return "\(delta / 60)m ago" }
        if delta < 86_400     { return "\(delta / 3600)h ago" }
        if delta < 2 * 86_400 { return "yesterday" }
        if delta < 7 * 86_400 { return "\(delta / 86_400) days ago" }
        let fmt = DateFormatter()
        fmt.dateFormat = "MMM d"
        return fmt.string(from: date)
    }
}

private extension String {
    var nonEmpty: String? {
        trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? nil : self
    }
}
